pinot-common/src/main/java/com/linkedin/pinot/common/utils/time/DateTimeFieldSpecUtils.java

```diff
@@ -0,0 +1,393 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.common.utils.time;
+
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.lang3.EnumUtils;
+import org.codehaus.jackson.annotate.JsonIgnore;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.Months;
+import org.joda.time.MutableDateTime;
+import org.joda.time.Weeks;
+import org.joda.time.Years;
+import org.joda.time.format.DateTimeFormat;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.linkedin.pinot.common.data.DateTimeFieldSpec;
+import com.linkedin.pinot.common.data.DateTimeFieldSpec.TimeFormat;
+
+public class DateTimeFieldSpecUtils {
+
+  private static final String FORMAT_TOKENS_ERROR_STR = "format must be of pattern size:timeunit:timeformat(:pattern)";
+  private static final String FORMAT_PATTERN_ERROR_STR = "format must be of format [0-9]+:<TimeUnit>:<TimeFormat>(:pattern)";
+  private static final String TIME_FORMAT_ERROR_STR =
+      "format must be of format [0-9]+:<TimeUnit>:EPOCH or [0-9]+:<TimeUnit>:SIMPLE_DATE_FORMAT:<format>";
+  private static final String GRANULARITY_TOKENS_ERROR_STR = "granularity must be of format size:timeunit";
+  private static final String GRANULARITY_PATTERN_ERROR_STR = "granularity must be of format [0-9]+:<TimeUnit>";
+  private static final String NUMBER_REGEX = "[1-9][0-9]*";
+
+  public static final String COLON_SEPARATOR = ":";
+
+  /* DateTimeFieldSpec format is of format size:timeUnit:timeformat:pattern(if sdf) */
+  private static final int FORMAT_SIZE_POSITION = 0;
+  private static final int FORMAT_UNIT_POSITION = 1;
+  private static final int FORMAT_TIMEFORMAT_POSITION = 2;
+  private static final int FORMAT_PATTERN_POSITION = 3;
+  private static final int MIN_FORMAT_TOKENS = 3;
+  private static final int MAX_FORMAT_TOKENS = 4;
+
+  /* DateTimeFieldSpec granularity is of format size:timeUnit */
+  private static final int GRANULARITY_SIZE_POSITION = 0;
+  private static final int GRANULARITY_UNIT_POSITION = 1;
+  private static final int MAX_GRANULARITY_TOKENS = 2;
+
+
+  private static final DateTime EPOCH_START_DATE = new DateTime(0L, DateTimeZone.UTC);
+
+  public enum CustomTimeUnit {
```

Comment:
The original implementation had Java and Custom TimeUnits, as they were added a different times. We should have a single time converter (probably JODA based) that handles all time units.

Reply:
Fixed